Stately 2.0 launch post (#270)
statelyai/docs: blog/2023-11-20-stately-studio-2-0/index.mdx

But Stately Studio has bigger ambitions than just being a suite of devtools for XState. We’ve frequently heard that these state diagrams are an important **source of truth** for critical app logic, serving as documentation for the entire team that stays up-to-date with your code. But a reliable source of truth for app logic is a need for **all** apps, not just those that use state machines directly.

That’s why we’re so excited to release Stately Studio 2.0, which aims to **meet developers where they are**, no matter which libraries, frameworks, or even languages they use. There are many benefits to modeling app logic with state diagrams and the actor model, and we want to enable developers to take advantage of those benefits to build more robust, feature-rich, and maintainable app logic faster.

<!--truncate-->

[image]

> Stately Studio 2.0 is an intelligent software modeling platform for working with app logic at all stages of the software development lifecycle. With its many specialized graph algorithms for analyzing state machines deterministically, paired with AI (large language models (LLMs) and more) for the not-so-deterministic tasks, such as generation & modification from natural language, [Stately Studio](https://stately.ai) is an essential tool for any development team working on complex app logic.

Reviewer: Really agree with this suggestion @gavination. I would add that the "and more" is awkward between the "(large language models)" and "to handle..." and doesn't add anything. It can just be removed.

Another option is to end with "Stately Studio is now and essential tool", to emphasize the "arrival" with this release, but if others disagree then I don't feel strongly about it.

> Stately Studio 2.0 is an intelligent software modeling platform for working with app logic at all stages of the software development lifecycle. With its many specialized graph algorithms for analyzing state machines deterministically, paired with AI (large language models (LLMs) and more) for the not-so-deterministic tasks, such as generation & modification from natural language, [Stately Studio](https://stately.ai) is an essential tool for any development team working on complex app logic.

Reviewer: Also, is there a reason to use the ampersand in "generation & modification" in this one location here as opposed to just "and"?

Let’s go over some of the entry actions to our ambitious transition.

## Import from XState, Redux, Zustand, and more

> In the front end, there are many popular state machine libraries, all with different focuses and tradeoffs. We’ve frequently said that developers are always writing state machines, whether they know it (explicit) or not (implicit). So, we want to make it easy to visualize app logic from any code as a state machine, starting with commonly used front-end state management libraries.

Reviewer: I also see "front-end" at the end of the paragraph already.

> In the front end, there are many popular state machine libraries, all with different focuses and tradeoffs. We’ve frequently said that developers are always writing state machines, whether they know it (explicit) or not (implicit). So, we want to make it easy to visualize app logic from any code as a state machine, starting with commonly used front-end state management libraries.

Author: I try to use `front end` as the noun, `front-end` as the adjective. And avoid `frontend` because it’s a weird hybrid word 😂

> In the front end, there are many popular state machine libraries, all with different focuses and tradeoffs. We’ve frequently said that developers are always writing state machines, whether they know it (explicit) or not (implicit). So, we want to make it easy to visualize app logic from any code as a state machine, starting with commonly used front-end state management libraries.

Author: @gavination I’m kind of reasoning on @davidkpiano’s part here, as he’s the original author of this bit. But I’d say that this is also the case outside of UI development when we’re talking about workflows etc?

> In the front end, there are many popular state machine libraries, all with different focuses and tradeoffs. We’ve frequently said that developers are always writing state machines, whether they know it (explicit) or not (implicit). So, we want to make it easy to visualize app logic from any code as a state machine, starting with commonly used front-end state management libraries.

Reviewer 2: So my understanding is that @davidkpiano was referencing UI dev specifically, since you're _always_  writing an implicit state machine with user interfaces. in regards to workflows development, it is technically possible to be dealing with scenarios where state machines aren't the answer, so I wouldn't say "always" in those cases.

- Easily detect logical errors in your app logic.
- Quickly generate test cases.

This feature is still experimental, but we’re excited for you to try it out. The conversion is made possible with a mixture of [AST](https://en.wikipedia.org/wiki/Abstract_syntax_tree) analysis and complex, state-machine-powered LLM logic. And this is just the start; we have many plans for features coming soon:

- Import *and* export to and from Redux, Zustand, and more libraries in the future. Of course, this has limitations; statecharts have many features beyond simple state management!
- Bidirectional editing of your code: edit the diagram, see the code update; edit the code, see the diagram update.
- Go further than libraries: import and export custom workflow [DSL]s and even other languages.

We also want to hear your ideas! [Request features and upvote suggestions](https://feedback.stately.ai) on our feedback board.

:::info

You can now get three free generations for our features using Stately AI. [Try them out today and upgrade to a Pro account](https://stately.ai/pricing) to get up to 1000 generations a month.

:::

## Generate UI prototypes

> Modeling flows as state machines for apps is useful, and bringing those state machines to life can make those flows more understandable. That is why you can now generate React user interfaces from the state diagrams you create (or generate) in Stately Studio.

Reviewer: ```suggestion
Modeling flows as state machines for apps is useful, and bringing those state machines to life can make flows more understandable. That is why you can now generate React user interfaces from the state diagrams you create (or generate) in Stately Studio.
```

[image]

With Stately AI and Stately Sky together, you can go from an idea to a deployed workflow in minutes. The Stately editor helps you create much more flexible workflows than other workflow tools—you can use these workflows on the front end, back end, or both. Perfect for onboarding flows, collaborative apps, games, AI agents, and more. Stately Sky is also backed by Cloudflare Workers so that you can run your state machines in 200+ locations worldwide, with low latency and high availability. 

You can [read more about Stately Sky in Gavin’s blog post](https://stately.ai/blog/2023-11-13-introducing-stately-sky).

## Enhanced Stately editor

Among all these huge new features, we have updated the Stately editor’s UI. We want [Stately.ai](http://stately.ai/) to feel as natural as a white-boarding or diagramming tool, while giving you the power to create robust state machines at any level of complexity.

> These updates significantly improve the editor’s performance, especially when working with large machines. We’ve also streamlined the editor with reorganized panels and a redesigned header and footer. The new design makes accessing the tools you need faster while also providing information about your machine, including its locked status, project visibility, and current version.

Reviewer: Any mention of new quick-access canvas buttons?

### Targetless transitions

> As a part of our improved support for XState in the Stately editor, the editor now supports targetless transitions. Targetless transitions are useful when you want a transition on your root state but don’t want to keep resetting your machine to the initial state. They also open up the possibility for multiple target state support in the future, which is handy if you want to transition to simultaneous parallel states.

Reviewer: Should we mention something about being able to trigger actions without any actual state change at all?

### Zoom to selection

You can now **zoom to selected** to re-center the canvas from the zoom menu or the right-click menu. This zoom is particularly helpful when working with large machines and demoing your machines to others, enabling you to jump to a different canvas area swiftly.

### Generate summary

As part of our Stately AI toolset for our Pro users, we now have an experimental **Get summary** option in the machine details panel. You can use this to create a summary of your state machine to make it even more understandable to non-technical stakeholders. We’ve also found it helpful in generating alternative text for images of our state machines. This feature could explain and summarize specific paths throughout the machine in the future, representing different use cases or edge cases.

[image]

## What’s coming next

And soon, there will be even more! We have designed many updates to the Stately editor to support even more capabilities that are on the way. Including:

- GitHub Sync
- Inspection capabilities
- System views and sequence diagrams for actors
- Stately Sky analytics
- XState version 5

> - Updated VSC Code extension

Reviewer: Is it the right time to mention something to do with multi-player or at least multi-presence in the Editor?

> ## Intelligent software modeling is the future

Reviewer: This last section is a bit long. Visually, it looks like a lot of bland text that readers might be tempted to just skip. The first sections are more of the summary but maybe the last 2 paragraphs can be in their own section like "Talk to us" or something more CTA.

Author: You’re right. I’ve pulled out a paragraph that’s bit repetitive of the introduction to hopefully tighten this up.

Right now, everyone is talking about how AI can help us work faster and smarter. LLMs can significantly accelerate state machine creation, modification, optimization, and explanation. Likewise, state machines can serve as essential guardrails for LLMs, helping them to avoid making mistakes by setting predetermined paths (state transitions) for them to follow. We even have open-source tooling that enables these guardrails for all developers, and we’ll release it soon.

We’re also using LLMs to help power the AI in [Stately Studio](http://stately.ai/), but our core technology and the end result are still good old-fashioned state machines, statecharts, and the actor model.

Our mission at Stately is to empower every software development team with intelligent visual tools that help you create, iterate on, test, diagram, document, analyze, and even deploy the various flows in your application, no matter how complex they may get. And importantly, we want to meet developers where they’re at by enabling this tooling for any codebase and at any level of adoption.

> If you have any ideas for features that would benefit your team, [share your feature request on our feedback board](https://feedback.stately.ai). We want your application logic to be understood by the entire team and always up-to-date, and we want to make that as easy as possible.

Reviewer: ```suggestion
If you have any ideas for features that would benefit your team, [share your feature request on our feedback board](https://feedback.stately.ai). We want to make it as easy as possible for your application logic to be understood by your entire team and always up-to-date.
```

Author: I’ve tweaked this slightly differently.

<p>
  <ThemedImage
    alt="Split view with state machine for a sensor reading app on the left, and a generated UI for the sensor readings on the right. The generated UI has a title, the current sensor readings, and buttons to log new readings. On top of the split view is a Generate React app button."
    sources={{
      light:
        '/blog/2023-11-20-stately-studio-2-0/generate.png',
      dark: '/blog/2023-11-20-stately-studio-2-0/generate-dm.png',
    }}
  />
</p>

You can open the React app in CodeSandbox to view the working React app built on top of your state machine. Whether you’re making a working prototype or creating a baseline for a functioning app, we want to make it easy to generate UI code from your state machines. We’ll also support StackBlitz in the future.

Of course, if you want to make the React prototype a little more lively, you can use our new **Enhance UI** button, which will use Stately AI to fill in more UI details. Enhance UI is an experimental feature, so it may not give perfect results, but it should provide a great starting point you can easily use for prototyping or even building out full apps.

## Drag-and-drop state assets

> A source of truth doesn’t really tell the whole truth if it’s just text, especially for applications with user interfaces. For a long time, our states have had descriptions that support markdown, including images from URLs, but we wanted to make that easier. Our Pro users can now drag and drop images directly onto states, and they will be uploaded and displayed in the state. You can adjust the image size and even upload multiple assets to a single state, arranging which asset is shown on the canvas from the state details panel.

Reviewer 3: ```suggestion
A source of truth doesn’t really tell the whole truth if it’s just text, especially for applications with user interfaces. For a long time, our states have had descriptions that support markdown, including images from URLs, but we wanted to make that easier. If you're a Stately Pro user, you can now drag and drop images directly onto states, and they will be uploaded and displayed in the state. You can adjust the image size and even upload multiple assets to a single state, arranging which asset is shown on the canvas from the state details panel.
```

<p>
  <ThemedImage
    alt="Our room readings state machine with the machine details panel open. Below a Get Summary button is the summary text for this machine that reads 'This state machine is called Accumulate room readings. Its purpose is to collect temperature and humidity readings from loT sensors and generate a report every hour. The state machine starts in the ConsumeReadings state, where it initializes the temperature and humidity values as null. It then waits for temperature and humidity events to be logged. Whenever a temperature or humidity event is received, the state machine updates the respective value. After one hour (3600000 milliseconds), the state machine transitions to the GenerateReport state. In this state, it invokes a separate machine called produceReport to generate the report/ Once the report is generated, it…'"
    sources={{
      light:
        '/blog/2023-11-20-stately-studio-2-0/summary.png',
      dark: '/blog/2023-11-20-stately-studio-2-0/summary-dm.png',
    }}
  />
</p>

## Pricing updates

> We’re also simplifying our pricing model. Individual developers can get a Stately Pro subscription for $29/month or $299/year. Teams can get a Stately Team subscription for $199/month or $1999/year for up to 10 users, which includes the full Stately Pro features as well as collaborative features for everyone on the team. We also have an Enterprise plan for larger teams and organizations. You can [read more about our pricing plans here](https://stately.ai/pricing).

Author: ```suggestion
We’re also simplifying our pricing model. Individual developers can get a Stately Pro subscription for $29/month or $299/year. Teams can get a Stately Team subscription for $199/month or $1999/year for up to 10 users. Team subscriptions includes collaborative features as well as the full Stately Pro features for everyone on the team. We also have an Enterprise plan for larger teams and organizations. [Read more about our pricing plans](https://stately.ai/pricing).
```
